engines/sci/graphics/paint16.cpp

```diff
@@ -131,31 +153,41 @@ void GfxPaint16::drawCel(GfxView *view, int16 loopNo, int16 celNo, const Common:
 		view->drawScaled(celRect, clipRect, clipRectTranslated, loopNo, celNo, priority, scaleX, scaleY, scaleSignal);
 }
 
-// This is used as replacement for drawCelAndShow() when hires-cels are drawn to
-// screen. Hires-cels are available only SCI 1.1+.
-void GfxPaint16::drawHiresCelAndShow(GuiResourceId viewId, int16 loopNo, int16 celNo, uint16 leftPos, uint16 topPos, byte priority, uint16 paletteNo, reg_t upscaledHiresHandle, uint16 scaleX, uint16 scaleY) {
+// This is used as replacement for drawCelAndShow() when hires-cels are drawn to screen. Hires-cels are available only SCI 1.1+.
+void GfxPaint16::drawHiresCelAndShow(GuiResourceId viewId, int16 loopNo, int16 celNo, uint16 leftPos, uint16 topPos, byte priority, uint16 paletteNo, reg_t hiresHandle, bool storeDrawingInfo) {
 	GfxView *view = _cache->getView(viewId);
 	if (!view)
 		return;
 
-	byte *memoryPtr = _segMan->getHunkPointer(upscaledHiresHandle);
-	if (!memoryPtr)
-		error("drawHiresCelAndShow: Invalid hires handle");
-
-	Common::Rect upscaledHiresRect;
-	_screen->bitsGetRect(memoryPtr, &upscaledHiresRect);
-	Common::Point topLeft(upscaledHiresRect.left, upscaledHiresRect.top);
-	Common::Point bottomRight(upscaledHiresRect.right, upscaledHiresRect.bottom);
-
-	topLeft = _screen->gfxDriver()->getRealCoords(topLeft);
-	bottomRight = _screen->gfxDriver()->getRealCoords(bottomRight);
+	byte *memoryPtr = _segMan->getHunkPointer(hiresHandle);
+	if (!memoryPtr) {
+		// The original KQ6WinCD interpreter does not treat this as an error, it just skips the hires drawing. It happens all the time
+		// when attempting to redraw the hires cels from the _hiresDrawObjs chain, we're supposed to just skip the invalidated items.
+		return;
+	}
 
+	Common::Rect portraitRect;
+	_screen->bitsGetRect(memoryPtr, &portraitRect);
+	Common::Rect clipRect(makeHiresRect(portraitRect));
 	Common::Rect celRect(view->getWidth(loopNo, celNo), view->getHeight(loopNo, celNo));
-	Common::Rect clipRect(topLeft.x, topLeft.y, bottomRight.x, bottomRight.y);
-	celRect.translate(leftPos + topLeft.x, topPos + topLeft.y);
+	celRect.translate(leftPos + clipRect.left, topPos + clipRect.top);
 	clipRect.clip(celRect);
 
 	view->draw(celRect, clipRect, clipRect, loopNo, celNo, priority, paletteNo, true);
+
+	// The original KQ6WinCD interpreter saves the hires drawing information in a linked list. There are two use cases, one is redrawing the
+	// window background when recieving WM_PAINT messages (which is irrelevant for us, since that happens in the backend) and the other is
```

Comment:
```suggestion
	// window background when receiving WM_PAINT messages (which is irrelevant for us, since that happens in the backend) and the other is
```